tests/quick.rs

```diff
@@ -0,0 +1,200 @@
+extern crate evmap;
+
+use evmap::{ReadHandle, WriteHandle};
+
+extern crate quickcheck;
+#[macro_use(quickcheck)]
+extern crate quickcheck_macros;
+
+use quickcheck::Arbitrary;
+use quickcheck::Gen;
+
+use rand::Rng;
+use std::cmp::{min, Ord};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::{BuildHasher, Hash};
+use std::ops::Deref;
+
+fn set<'a, T: 'a, I>(iter: I) -> HashSet<T>
+where
+    I: IntoIterator<Item = &'a T>,
+    T: Copy + Hash + Eq,
+{
+    iter.into_iter().cloned().collect()
+}
+
+#[quickcheck]
+fn contains(insert: Vec<u32>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+
+    insert.iter().all(|&key| r.get(&key).is_some())
+}
+
+#[quickcheck]
+fn contains_not(insert: Vec<u8>, not: Vec<u8>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+
+    let nots = &set(&not) - &set(&insert);
+    nots.iter().all(|&key| r.get(&key).is_none())
+}
+
+#[quickcheck]
+fn insert_empty(insert: Vec<u8>, remove: Vec<u8>) -> bool {
+    let (r, mut w) = evmap::new();
+    for &key in &insert {
+        w.insert(key, ());
+    }
+    w.refresh();
+    for &key in &remove {
+        w.empty(key);
+    }
+    w.refresh();
+    let elements = &set(&insert) - &set(&remove);
+    r.len() == elements.len()
+        && r.map_into::<_, Vec<()>, _>(|_, _| ()).len() == elements.len()
```

Comment:
I've been using `map_into` a lot. I realized that I could use `r.read().iter().flatten()`. I suppose that's clearer becase we can ditch a lot of the generic typing. But I wonder if it might be worth implementing `iter` or `keys` or something for the `ReadHandle`?

Reply:
Isn't `IntoIter` already implemented for `&ReadHandle`?

Reply:
Hmmmm maybe I'm just not seeing it? I'm only seeing iter methods on `MapReadRef` and `Values` (and `MapReadRef` does have an `into_iter` but it's the same as `iter` which is a little surprising 🤔).

Reply:
Ohh, wait, sorry, I misread `ReadHandle`. I _don't_ think we want to impl `iter()` directly on `ReadHandle`, because it would make the call to `read()` implicit. That call holds up writers while it's active, so I'd prefer to have that call be as visible as possible.

Reply:
Ahhh that makes oodles of sense. Which is why `map_into` is fine because it creates the iterator but consumes it immediately.

Reply:
And we couldn't have a `keys` helper unless we require `K: Clone` or `K: Copy` because we can't give out a `Vec<&K>`/`Iterator<Item=&K>`, right? Or should we add an `impl` block for that case? It's probably only helpful for tests or other single-threaded applications in which case you probably wouldn't reach for this tool?

Reply:
I think `keys` and `values` can (and probably should) exist on `MapReadRef`. I don't think we want to add any more methods to `ReadHandle` that take the guard behind the scenes. Even `map_into` is _arguably_ a foot-gun.

Reply:
👍 Roger that. I'd be glad to do that in a follow up PR! Don't want to scope creep this too much. And if we think these tests read all crappy then I can do a PR first for `keys` and `values` and then rebase this on top of that?

Reply:
I think they're probably fine as-is, and then we can tidy them up with the follow-on PR?

Reply:
Works for me!